content/2017.02.handling-python-projects-dependencies.rst

```diff
@@ -0,0 +1,392 @@
+that for you.
+
+Our take on this was to put test dependencies in a
+``dev-requirements.txt`` file.
+
+We have a Makefile rule that knows if dev-dependencies should be
+installed or not before running the ``tests`` target.
+
+As a bonus it will automatically create a virtualenv if you don't
+have one already activated:
+
+.. code-block:: Makefile
+
+    VIRTUALENV = virtualenv --python=python3
+    VENV := $(shell echo $${VIRTUAL_ENV-.venv})  # Use the activated virtualenv path or use .venv
+    PYTHON = $(VENV)/bin/python
+    INSTALL_STAMP = $(VENV)/.install.stamp
+    INSTALL_DEV_STAMP = $(VENV)/.dev_env_installed.stamp
+    
+    install: $(INSTALL_STAMP)
+    
+    $(INSTALL_STAMP): $(PYTHON) setup.py  # Refresh the virtualenv if setup.py changed
+    	$(VENV)/bin/pip install -U pip
+    	$(VENV)/bin/pip install -Ue .
+    	touch $(INSTALL_STAMP)
+    
+    virtualenv: $(PYTHON)  # Create the virtualenv if needed (python executable not present)
+    
+    $(PYTHON):
+    	$(VIRTUALENV) $(VENV)
+    	
+    install-dev: install $(DEV_STAMP)
+    
+    # Refresh dev dependencies if dev-requirements.txt changed
+    $(DEV_STAMP): $(PYTHON) dev-requirements.txt
+    	$(VENV)/bin/pip install -Ur dev-requirements.txt
+    	touch $(DEV_STAMP)
+    
+    tests-once: install-dev
+        $(VENV)/bin/py.test tests/
+    
+    serve: install
+    	$(VENV)/bin/python manage.py runserver
+
+
+Conclusion
+==========
+
+As a conclusion, when working with Python dependencies there are three
+process that needs to work well together.
+
+- Installing automatically a project version in a working state,
+  that's what you can do using requiremrents constraint files.
+- Keeping the project up-to-date with next version of its dependencies
+  without breaking previous released versions.
+- Installing a project as part as another project without it breaking
+  the project with dependencies version conflicts.
+
+Our solution to this is as follow:
+
+- Keep your project dependencies clean of any version number.
+- At release time document what is the known version check.
+- Use that known working state as a way to install a given release of
+  the project.
+- Run your CI without the constraint file to detect dependencies
+  update that might break future release of your code.
+
+
+In our way to the future
+========================
+
+In a perfect world, it would be great to know in advance what is the
+next version of the library that will stop working with your code.
+
+The good news is that this future is already there and that's what
+semantic versionning is trying to address.
+
+
+What can break your code?
+-------------------------
+
+- Changing the name of functions or objects
+- Changing the name of parameters or their order in the function call
+- Changing the way to configure the project
+
+In one word, everything that changes the API that is exposed by the
+library will break the code of people relying on it.
+
+
+What is Semantic Versionning?
```

Comment:
`Versioning` (here and everywhere). Double-N is the exception, it only happens to differentiate verbs that end in `-n` from those that end in `-ne`. (`pin` -> `pinning`, `pine` -> `pining`) Most of these verbs are very short.